src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -124,6 +127,72 @@ public void updateState(ChannelState state) {
         }
     }
 
+    @Override
+    public PusherEvent prepareEvent(String event, String message) {
+        try {
+
+            Map receivedMessage = GSON.fromJson(message, Map.class);
+            final String decryptedMessage = decryptMessage((String) receivedMessage.get("data"));
+            receivedMessage.replace("data", decryptedMessage);
+
+            return GSON.fromJson(
+                    GSON.toJson(receivedMessage), PusherEvent.class);
+
+        } catch (AuthenticityException e1) {
+
+            // retry once only.
+            disposeSecretBoxOpener();
+            authenticate();
+
+            try {
+                Map receivedMessage = GSON.fromJson(message, Map.class);
+                final String decryptedMessage = decryptMessage((String) receivedMessage.get("data"));
+                receivedMessage.replace("data", decryptedMessage);
+
+                return GSON.fromJson(
+                        GSON.toJson(receivedMessage), PusherEvent.class);
+            } catch (AuthenticityException e2) {
+                disposeSecretBoxOpener();
```

Comment:
What will happen to the next event if we dispose of the secret here on the final attempt?

Reply:
That's a good question. It will probably crash with an exception as the secretbox will be null. What should we be doing in this situation? We can do a check to see if the secretbox is null at the start of the prepareEvent, and call the `onDecryptionFailure` with a different reason?

At any point, there should definitely be a test for this which i will add now.

Reply:
Have added the test https://github.com/pusher/pusher-websocket-java/pull/249/commits/4bfc70de174c9abae8c6a012e29a2507c3570fd9#diff-adf187cd707fe9cbe8285aefab4bafbbR287 - and assumed i was okay to implement the logic as described. Let me know what you think on this.

Reply:
There are circumstances during the rotation of master key on the customer's backend where races can occur between events sent and keys received, so I think it's important that the SDK recover from a failed decryption and not fail all future events because of a period of instability when the keys are rotated.

Personally, I would leave the last fetched key in place and attempt the next event using it.

Reply:
👍 Marek left feedback before that he expected it to be cleared after each failed decryption.

I think what you say makes sense - each subsequent message should be able to send 1 retry to the authorization endpoint to get a new shared_secret. 

I've updated the code to do this.

I've also updated the tests: there is one test to ensure that two messages that have incorrect shared secrets after a retry, i also have a test for if a second message when it does the retry gets a correct key, it should succeed. 

Does that make sense?

Reply:
Yup, sounds great